aws/sns/resources.py

```diff
@@ -21,20 +22,49 @@ def sns_topics():
     )
 
 
-def sns_subscription_attributes():
+def sns_parent_topic_exists(subscription):
     "https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#subscription"
-    return [
+    topicArn = ":".join(
+        re.split(":", subscription)[0:6]
+    )  # Naive method of divining topic from subscription
+    try:
         botocore_client.get(
             service_name="sns",
-            method_name="get_subscription_attributes",
+            method_name="get_topic_attributes",
             call_args=[],
-            call_kwargs={"SubscriptionArn": subscription["SubscriptionArn"]},
-            profiles=[subscription["__pytest_meta"]["profile"]],
-            regions=[subscription["__pytest_meta"]["region"]],
+            call_kwargs={"TopicArn": topicArn},
+        )
+        return True
+    except:
+        return False
+
+
+def sns_get_subscription_attrs(subscription):
+    "https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#subscription"
+    # This function catches the error thrown by the AWS API when attempting to query the attributes of
+    # a subscription that has no parent topic.
+    try:
+        attrs = (
+            botocore_client.get(
+                service_name="sns",
+                method_name="get_subscription_attributes",
+                call_args=[],
+                call_kwargs={"SubscriptionArn": subscription["SubscriptionArn"]},
+                profiles=[subscription["__pytest_meta"]["profile"]],
+                regions=[subscription["__pytest_meta"]["region"]],
+            )
+            .extract_key("Attributes")
+            .values()[0]
         )
-        .extract_key("Attributes")
-        .values()[0]
-        for subscription in sns_subscriptions()
+        return attrs
+    except:
```

Comment:
A more specific `except` clause would also be nice here.

Reply:
Do you have any suggestions on what to add?  I'm a bit out of my depth here.

Reply:
There's a specific error you want to avoid here, right? The current code basically says "whatever goes wrong, just paper over it and carry on." What I believe you actually want is "I know the API call above sometimes throws this specific error, and you don't need to mind that one."

So if you can reproduce this error, or have any execution log of that error occurring, you can see what the error class of that error was `botocore.exceptions.ClientError`, then you can import that error class into this module and do

    except botocore.exceptions.ClientError:

here. I don't know what the right error class is – that depends on the errors you have encountered here.

This has two advantages over the current code. First, it makes it clearer what error you want to ignore, which helps readers of the code. (Code is read much more often than it is written.) And second, it doesn't paper over some random other errors you didn't know could occur here, so you will still see unexpected errors in the logs.

In some cases you may want to be even more specific than restricing the errors you catch to a specific class. In these cases you can inspect the exception object an re-raise it isn't what you expected:

    except GlonkError as exc:
        if exc.code != 2345:        # completely made up – custom exceptions
                                    # usually have fields with additional info
            raise                   # reraise exception if it doesn't have the right code

See also [this StackOverflow question](https://stackoverflow.com/questions/46174385/properly-catch-boto3-errors) for a more on-topic example.

Or you can be less specific, by catching multiple classes with `except TypeError, ValueError:` or base classes of exceptions, which will match all exceptions derived from that base class. Unless you have specific requirements, I'd keep it simple at first and just catch the exception by class.

You almost never want a bare `except:` class without any further restriction. This will also catch `KeyboardInterrupt` and `SystemExit` and just pretend it didn't happen, which probably isn't what you meant. To broadly catch all application exceptions, you can use `except Exception:`; `Exception` is the base class of most built-in exceptions and virtually all user-defined exceptions, so this is almost always better than just `except:`.

(Sorry, this was probably more than you asked for – I somehow switched to StackOverflow mode. :-) )

Reply:
It may have been more than I _asked_ for, but it was exactly what I needed.  Thank you for the very detailed response.  I believe this latest commit sufficiently limits the exception handling to the specific error condition we want to trap.